### PR TITLE
feat: support strategic merge patch in overrides

### LINKED ISSUE #694: Need to be able to override/replace livenessProbe handler
The default livenessProbe for a service looks like: 
```
livenessProbe:
      failureThreshold: 3
      initialDelaySeconds: 30
      periodSeconds: 10
      successThreshold: 1
      tcpSocket:
        port: rpc
      timeoutSeconds: 1
```
We are able to override all the fields, except for the handler. For example, folks should be able to add overrides like below:
```
livenessProbe:
          grpc:
                    port: 7233
                    service: frontend.temporal.temporal.svc.cluster.local
```
or 
```
livenessProbe:
          exec:
                    command: [
                              "/etc/temporal/grpc-health-probe/grpc_health_probe",
                              "-addr=:7233"
                              ...
                    ]
```

Unfortunately, right now this results in error : `Deployment.extensions "temporal-frontend" is invalid: spec.template.spec.containers[0].livenessProbe.tcpSocket: Forbidden: may not specify more than 1 handler type` as the operator attempts to add both handlers.

### PR DESCRIPTION
**Pull request by yujunz**
Close #694

### CODE REVIEW

**Comment by alexandrevilain** [comment]
Thanks for this contribution @yujunz !

Is there any breaking change on the API ? Could we still use YAML as done currently in overrides ?

I need some time to deeply test this.

Maybe the following hack could be removed: https://github.com/alexandrevilain/temporal-operator/blob/main/Makefile#L49 ?

**Comment by yujunz** [comment]
Theoretically no breaking changes since `JSON` is a superset of `PodSpec`. All valid `PodSpec` in override should work as before.

One known change is that overrides no longer unmarshal to `PodSpec` schema, but invalid or unknown fields should get rejected or ignored when applying. All unit tests pass but not sure how to make e2e testing.

File: pkg/kubernetes/overrides_test.go [hunk 1]
```diff
@@ -276,19 +318,65 @@ func TestApplyDeploymentOverrides(t *testing.T) {
...
 							},
+						},
+					},
+				},
+			},
+			override: &v1beta1.DeploymentOverride{
+				Spec: &v1beta1.DeploymentOverrideSpec{
+					Template: &v1beta1.PodTemplateSpecOverride{
+						Spec: &apiextensionsv1.JSON{
+							Raw: []byte(`{"containers":[{"name":"test","livenessProbe":{"$patch":"replace","tcpSocket":null,"exec":{"command":["echo","hi"]}}}]}`),
```
Comment (guritinvda) [inline pkg/kubernetes/overrides_test.go:341]: You shouldn't need to specify "tcpSocket":null. Right?

Comment (yujunz) [reply pkg/kubernetes/overrides_test.go:341]: It turns out we have too. `replace` applies to every sibling keys, e.g. `tcpSocket`, `exec` in this case. `tcpSocket` will remain as is if not set explicitly to `null`. I also tried to replace `livenessProbe` but that seems removing other keys like delay seconds.

File: pkg/kubernetes/overrides_test.go [hunk 2]
```diff
@@ -130,18 +132,21 @@ func TestApplyDeploymentOverrides(t *testing.T) {
...
-										Limits: corev1.ResourceList{
-											corev1.ResourceCPU: apimachineryresource.MustParse("100m"),
-										},
-									},
-								},
-							},
+						Spec: &apiextensionsv1.JSON{
+							Raw: []byte(`{"containers":[{"name":"testcontainer","resources":{"limits":{"cpu":"100m"}}}]}`),
 						},
+						// Spec: &corev1.PodSpec{
```
Comment (alexandrevilain) [inline pkg/kubernetes/overrides_test.go:138]: Could you please remove all commented parts ?

**Review by alexandrevilain — CHANGES_REQUESTED** [review]
IMO the Makefile could be cleaned up now, am I wrong ?
Could you also update the documentation to add the liveness probe example ?

File: pkg/kubernetes/overrides_test.go (hunk 2 above)
Comment (yujunz) [reply pkg/kubernetes/overrides_test.go:138]: Done

**Comment by yujunz** [comment]
> IMO the Makefile could be cleaned up now, am I wrong ?
Not very clear about the existing hacks in Makefile. Could you please suggest the change?

> Could you also update the documentation to add the liveness probe example ?
Done

File: pkg/kubernetes/overrides_test.go (hunk 2 above)
Comment (yujunz) [reply pkg/kubernetes/overrides_test.go]: The json was actually generated by GitHub Copilot from the commented code 😊.

**Comment by yujunz** [comment]
All comments addressed. @alexandrevilain

**Review by alexandrevilain — APPROVED** [review]

**Pull request merged**